classfile/reader.ts

```diff
@@ -0,0 +1,121 @@
+module J2ME {
+  declare var util;
+  export class Reader {
+    view: DataView;
+    // DataView is not optimized, use Uint8Array for the fast paths.
+    u8: Uint8Array;
+    offset: number;
+    static T: string [] = new Array(124);
+    constructor(buffer: ArrayBuffer, offset: number) {
+      this.view = new DataView(buffer);
+      this.u8 = new Uint8Array(buffer);
+      this.offset = offset || 0;
+    }
+
+    read8() {
+      return this.u8[this.offset++];
+    }
+
+    read16() {
+      var u8 = this.u8;
+      var o = this.offset;
+      this.offset += 2;
+      return u8[o] << 8 | u8[o + 1];
+    }
+
+    read32() {
+      return this.readInteger() >>> 0;
+    }
+
+    readInteger() {
+      var o = this.offset;
+      var u8 = this.u8;
+      var a = u8[o + 0];
+      var b = u8[o + 1];
+      var c = u8[o + 2];
+      var d = u8[o + 3];
+      this.offset = o + 4;
+      return (a << 24) | (b << 16) | (c << 8) | d;
+    }
+
+    readFloat() {
+      var data = this.view.getFloat32(this.offset, false);
+      this.offset += 4;
+      return data;
+    }
+
+    readDouble() {
+      var data = this.view.getFloat64(this.offset, false);
+      this.offset += 8;
+      return data;
+    }
+
+    readStringFast(length: number): string {
+      var r = Reader.T;
+      r.length = length;
+      var i = 0, j = 0;
+      var o = this.offset;
+      var e = o + length;
+      var u8 = this.u8;
+      while (o < e) {
+        var x = u8[o++];
+        if (x <= 0x7f) {
+          // Code points in the range '\u0001' to '\u007F' are represented by a
+          // single byte.
+          // The 7 bits of data in the byte give the value of the code point
+          // represented.
+          r[j++] = String.fromCharCode(x);
+        } else if (x <= 0xdf) {
+          // The null code point ('\u0000') and code points in the range '\u0080'
+          // to '\u07FF' are represented by a pair of bytes x and y.
+          var y = u8[o++]
+          r[j++] = String.fromCharCode(((x & 0x1f) << 6) + (y & 0x3f));
+        } else {
+          // Code points in the range '\u0800' to '\uFFFF' are represented by 3
+          // bytes x, y, and z.
+          var y = u8[o++];
+          var z = u8[o++];
+          r[j++] = String.fromCharCode(((x & 0xf) << 12) + ((y & 0x3f) << 6) + (z & 0x3f));
+        }
+      }
+      this.offset = o;
+      r.length = j;
+      return r.join("");
+    }
+
+    readString(length) {
+      if (length === 1) {
+        var c = this.u8[this.offset];
+        if (c <= 0x7f) {
+          this.offset ++;
+          return String.fromCharCode(c);
+        }
+      } else if (length < 1024) {
+        return this.readStringFast(length);
+      }
+      return this.readStringSlow(length);
```

Comment:
I was testing in the shell where we apparently use a polyfill because `TextDecoder` is not available. Measuring in the browser is a different story. The sweet spot there is around 40 bytes (for strings that don't contain 2 or 3 byte code points.)

http://jsbin.com/sefiyubose/1/edit?html,js,output

```
Read: 0 bytes, x faster in JS vs TextDecoder: 2.12 (3.58 / 1.69)
Read: 1 bytes, x faster in JS vs TextDecoder: 4.79 (5.42 / 1.13)
Read: 2 bytes, x faster in JS vs TextDecoder: 6.04 (3.23 / 0.53)
Read: 3 bytes, x faster in JS vs TextDecoder: 4.59 (2.66 / 0.58)
Read: 4 bytes, x faster in JS vs TextDecoder: 4.44 (2.89 / 0.65)
Read: 5 bytes, x faster in JS vs TextDecoder: 3.55 (2.50 / 0.70)
Read: 6 bytes, x faster in JS vs TextDecoder: 3.73 (2.86 / 0.77)
Read: 7 bytes, x faster in JS vs TextDecoder: 3.29 (2.64 / 0.80)
Read: 8 bytes, x faster in JS vs TextDecoder: 5.09 (4.95 / 0.97)
Read: 9 bytes, x faster in JS vs TextDecoder: 2.91 (2.66 / 0.91)
Read: 10 bytes, x faster in JS vs TextDecoder: 2.53 (2.53 / 1.00)
Read: 11 bytes, x faster in JS vs TextDecoder: 3.75 (3.92 / 1.05)
Read: 12 bytes, x faster in JS vs TextDecoder: 2.22 (2.65 / 1.19)
Read: 13 bytes, x faster in JS vs TextDecoder: 2.17 (2.69 / 1.24)
Read: 14 bytes, x faster in JS vs TextDecoder: 3.08 (4.12 / 1.34)
Read: 15 bytes, x faster in JS vs TextDecoder: 1.89 (2.63 / 1.39)
Read: 16 bytes, x faster in JS vs TextDecoder: 1.82 (2.54 / 1.40)
Read: 17 bytes, x faster in JS vs TextDecoder: 1.77 (2.57 / 1.45)
Read: 18 bytes, x faster in JS vs TextDecoder: 2.31 (3.72 / 1.61)
Read: 19 bytes, x faster in JS vs TextDecoder: 1.72 (2.59 / 1.50)
Read: 20 bytes, x faster in JS vs TextDecoder: 1.60 (2.64 / 1.65)
Read: 21 bytes, x faster in JS vs TextDecoder: 3.11 (5.43 / 1.74)
Read: 22 bytes, x faster in JS vs TextDecoder: 1.49 (3.52 / 2.36)
Read: 23 bytes, x faster in JS vs TextDecoder: 1.45 (2.89 / 2.00)
Read: 24 bytes, x faster in JS vs TextDecoder: 1.93 (4.29 / 2.23)
Read: 25 bytes, x faster in JS vs TextDecoder: 1.40 (3.12 / 2.24)
Read: 26 bytes, x faster in JS vs TextDecoder: 1.35 (3.10 / 2.29)
Read: 27 bytes, x faster in JS vs TextDecoder: 1.87 (4.27 / 2.29)
Read: 28 bytes, x faster in JS vs TextDecoder: 1.29 (3.17 / 2.47)
Read: 29 bytes, x faster in JS vs TextDecoder: 1.32 (3.22 / 2.43)
Read: 30 bytes, x faster in JS vs TextDecoder: 1.26 (3.17 / 2.52)
Read: 31 bytes, x faster in JS vs TextDecoder: 1.80 (4.54 / 2.53)
Read: 32 bytes, x faster in JS vs TextDecoder: 1.11 (3.01 / 2.70)
Read: 33 bytes, x faster in JS vs TextDecoder: 1.21 (3.02 / 2.50)
Read: 34 bytes, x faster in JS vs TextDecoder: 1.67 (4.48 / 2.67)
Read: 35 bytes, x faster in JS vs TextDecoder: 1.20 (3.26 / 2.71)
Read: 36 bytes, x faster in JS vs TextDecoder: 1.13 (3.13 / 2.77)
Read: 37 bytes, x faster in JS vs TextDecoder: 1.59 (4.45 / 2.80)
Read: 38 bytes, x faster in JS vs TextDecoder: 1.12 (3.20 / 2.85)
Read: 39 bytes, x faster in JS vs TextDecoder: 1.10 (3.36 / 3.06)
Read: 40 bytes, x faster in JS vs TextDecoder: 1.49 (4.41 / 2.96)
Read: 41 bytes, x faster in JS vs TextDecoder: 1.14 (3.40 / 2.99)
Read: 42 bytes, x faster in JS vs TextDecoder: 1.08 (3.40 / 3.15)
Read: 43 bytes, x faster in JS vs TextDecoder: 1.04 (3.32 / 3.18)
Read: 44 bytes, x faster in JS vs TextDecoder: 1.39 (4.54 / 3.27)
Read: 45 bytes, x faster in JS vs TextDecoder: 1.00 (3.21 / 3.21)
Read: 46 bytes, x faster in JS vs TextDecoder: 0.96 (3.40 / 3.53)
Read: 47 bytes, x faster in JS vs TextDecoder: 1.79 (6.20 / 3.46)
Read: 48 bytes, x faster in JS vs TextDecoder: 0.77 (3.67 / 4.74) 
```